### PR TITLE
Indirect - MSDFit Interface - Add Plot Guess feature

### LINKED ISSUE #20975: Indirect - MSDFit Interface - No Plot Guess feature
Currently in the Indirect > Data Analysis > MSDFit interface, there is no Plot Guess feature.

This feature should be added using the existing methods in IndirectDataAnalysisTab.

### PR DESCRIPTION
**Pull request by brandonhewer**
Currently in the MSDFit interface, there is no Plot Guess feature (as seen in the other Indirect Data Analysis Fit Interfaces).

This PR adds the plot guess feature to the MSDFit interface.

**To test:**
1. Navigate to Interfaces > Indirect > Data Analysis > MSDFit
2. Supply the provided file in the Input field.
3. For each model in the Model Input drop-down menu:
    1. Select the model in the Model Input drop-down menu.
    2. Check the Plot Guess checkbox.
    3. Ensure the guess is plotted (green plot) in the top preview plot.
    4. Move the blue fit selectors in the top preview plot and ensure the guess plot is updated.

**Must be merged after #20980**

**Test Files:**
[msdfit_test_file.zip](https://github.com/mantidproject/mantid/files/1402311/msdfit_test_file.zip)

Fixes #20975 . 

**Release Notes** 

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/scientific_interfaces/Indirect/ConvFit.cpp [hunk 1]
```diff
@@ -482,7 +481,7 @@ void ConvFit::algorithmComplete(bool error, const QString &outputWSName) {
   updatePlot();
   updatePlotRange();
 
-  std::string paramWsName = outputPrefix + "_Parameters_new";
+  std::string paramWsName = outputPrefix + "_Parameters";
```
Comment (gvardany) [inline qt/scientific_interfaces/Indirect/ConvFit.cpp]: const?

File: qt/scientific_interfaces/Indirect/ConvFit.cpp [hunk 2]
```diff
@@ -1293,7 +1205,8 @@ void ConvFit::updateProperties(int specNo) {
 
 void ConvFit::updateProperties(int specNo, const QString &fitFunction) {
   bool isTwoLorentzian = fitFunction == "Lorentzian 2";
-  bool specOutOfBounds = specNo < m_runMin || m_runMax < specNo;
+  bool specOutOfBounds =
```
Comment (gvardany) [inline qt/scientific_interfaces/Indirect/ConvFit.cpp]: const?

File: qt/scientific_interfaces/Indirect/ConvFit.cpp [hunk 3]
```diff
@@ -1344,61 +1257,17 @@ QVector<QString> ConvFit::indexToFitFunctions(const int &fitTypeIndex) {
...
-  FunctionDomain1DVector domain(dataX);
-  FunctionValues outputData(domain);
-  function->function(domain, outputData);
-
-  std::vector<double> dataY(nData);
-  for (size_t i = 0; i < nData; i++) {
-    dataY[i] = outputData.getCalculated(i);
+  if (m_uiForm.dsResInput->isValid() && m_uiForm.ckPlotGuess->isChecked()) {
+    extendResolutionWorkspace();
+    bool tieCentres = (m_uiForm.cbFitType->currentIndex() == 2);
```
Comment (gvardany) [inline qt/scientific_interfaces/Indirect/ConvFit.cpp]: const?

File: qt/scientific_interfaces/Indirect/IndirectDataAnalysisTab.cpp [hunk 4]
```diff
@@ -129,6 +165,257 @@ void IndirectDataAnalysisTab::plotCurrentPreview() {
   }
 }
 
+/**
+ * Plots the selected spectrum of the input workspace in this indirect data
+ * analysis tab.
```
Comment (gvardany) [inline qt/scientific_interfaces/Indirect/IndirectDataAnalysisTab.cpp]: @param ?

File: qt/scientific_interfaces/Indirect/IndirectDataAnalysisTab.cpp [hunk 5]
```diff
@@ -129,6 +165,257 @@ void IndirectDataAnalysisTab::plotCurrentPreview() {
...
+    auto workspace =
+        boost::dynamic_pointer_cast<MatrixWorkspace>(outputWS->getItem(index));
+    updatePlot(workspace, fitPreviewPlot, diffPreviewPlot);
+  } else {
+    diffPreviewPlot->clear();
+    plotInput(fitPreviewPlot);
+  }
+}
+
+void IndirectDataAnalysisTab::updatePlot(
```
Comment (gvardany) [inline qt/scientific_interfaces/Indirect/IndirectDataAnalysisTab.cpp]: missing oxygen for the methods hereafter

File: qt/scientific_interfaces/Indirect/IndirectDataAnalysisTab.cpp [hunk 6]
```diff
@@ -129,6 +165,257 @@ void IndirectDataAnalysisTab::plotCurrentPreview() {
...
+  const size_t binIndexLow = inputWS->binIndexOf(startX);
+  const size_t binIndexHigh = inputWS->binIndexOf(endX);
+  const size_t nData = binIndexHigh - binIndexLow;
+
+  const auto &xPoints = inputWS->points(0);
+
+  std::vector<double> dataX(nData);
+  std::copy(&xPoints[binIndexLow], &xPoints[binIndexLow + nData],
+            dataX.begin());
+  std::vector<double> dataY = computeOutput(func, dataX);
```
Comment (gvardany) [inline qt/scientific_interfaces/Indirect/IndirectDataAnalysisTab.cpp]: const

**Review by gvardany — CHANGES_REQUESTED** [review]
Tested with MSDFit, plot guess appears on the preview. Few minor oversights in the code, made some inline comments on some lines, but there are not exhaustive, so please look over:
- `const` consistency is not preserved
- doxygen documentation is missing for some methods

Also, these interfaces have a documentation page: http://docs.mantidproject.org/nightly/interfaces/Indirect%20Data%20Analysis.html#msd-fit 
One could add a statement about the new feature.

The main problem appears when adjusting the fitting range (the blue lines), when I brought them too close, I got the error attached, and there is no way out, even if I click continue. Maybe this can be handled in a more friendly manner?

screen shot 2017-11-03 at 10 46 33

**Comment by brandonhewer** [comment]
@gvardany Added a check for if the guess has enough data points before plotting - there is no reason to throw an error here - for a small region the guess just won't appear on the plot until you move the range selectors.

**Comment by gvardany** [comment]
That error is fixed, but when I try to bring the blue lines close after the guess is plotted, I get a similar one, how about this?

screen shot 2017-11-06 at 11 29 16

**Comment by brandonhewer** [comment]
@gvardany This error message should no longer appear.

**Review by gvardany — APPROVED** [review]

**Pull request merged**